packages/query/src/actions.ts

```diff
@@ -250,6 +257,89 @@ export async function getTypeData(
   return { params, fields };
 }
 
+/**
+ * Checks that `EXPLAIN EXECUTE` of the prepared statement works, otherwise returns the error.
+ * @param query query string, can only contain proper Postgres numeric placeholders
+ * @param typeData type data, the result from getTypeData for this same query
+ * @param queue
+ */
+export async function explainQuery(
+  query: string,
+  typeData: TypeData,
+  queue: AsyncQueue,
+): Promise<string[] | IParseError> {
+  if ('errorCode' in typeData) return typeData;
+  const uniqueName = crypto.createHash('md5').update(query).digest('hex');
+  // Prepare query
+  await queue.send(messages.parse, {
+    name: uniqueName,
+    query,
+    dataTypes: [],
+  });
+  await queue.send(messages.flush, {});
+  const parseResult = await queue.reply(
+    messages.errorResponse,
+    messages.parseComplete,
+  );
+  try {
+    if ('fields' in parseResult) {
+      // Error case
+      const { fields: errorFields } = parseResult;
+      return {
+        errorCode: errorFields.R,
+        hint: errorFields.H,
+        message: errorFields.M,
+        position: errorFields.P,
+      };
+    }
+
+    // Explain query (will throw error on permissions failure). Leverages the
+    // fact that nullability is not checked in types at this stage, so `null` is
+    // a valid value for all types (yes, even domain types with a `non null`
+    // constraint).
+    const length = typeData.params.length;
+    const params = Array.from({ length }, () => 'null');
+    const explain = await runQuery(
+      `explain execute ${escapeSqlIdentifier(uniqueName)}${
+        params.length === 0 ? '' : ` (${params.join(', ')})`
+      };`,
+      queue,
+    );
+
+    return explain.map((e) => e[0]);
+  } catch (e: any) {
+    // This is most likely from the `runQuery` statement
+    /* Example error:
+     * ```
+     * {
+     *   type: 'ServerError',
+     *   bufferOffset: 100,
+     *   severity: 'ERROR',
+     *   message: 'permission denied for table user_emails'
+     * }
+     * ```
+     */
+    console.error('Error occurred whilst testing query:', e);
+    return {
+      errorCode: '_ERROR',
+      message: e.message,
+    };
```

Comment:
This does not feel right, but it makes the error handled by `insert into user_emails(created_at)` be handled in the same way as `insert into user_emails(id)`, thereby making it so I can add the functioning tests. The former being an RBAC permission which is thrown at `EXPLAIN EXECUTE` time, whereas the latter is a generated column error which is thrown (I think) at parse time.